zerver/tests/test_soft_deactivation.py

```diff
@@ -90,3 +92,50 @@ def test_do_soft_activate_users(self) -> None:
         for user in users:
             user.refresh_from_db()
             self.assertFalse(user.long_term_idle)
+
+    def test_get_users_for_catch_up(self) -> None:
+        users = [
+            self.example_user('hamlet'),
+            self.example_user('iago'),
+            self.example_user('cordelia'),
+            self.example_user('ZOE'),
+            self.example_user('othello'),
+            self.example_user('prospero'),
+            self.example_user('aaron'),
+            self.example_user('polonius'),
+        ]
+        for user_profile in UserProfile.objects.all():
+            user_profile.long_term_idle = True
+            user_profile.save(update_fields=['long_term_idle'])
+
+        filter_kwargs = dict(realm=get_realm('zulip'))
+        users_to_catch_up = get_soft_deactivated_users_for_catch_up(filter_kwargs)
+
+        self.assert_length(users_to_catch_up, 8)
+        for user in users_to_catch_up:
+            self.assertTrue(user in users)
+
+    def test_do_catch_up_users(self) -> None:
+        stream = 'Verona'
+        hamlet = self.example_user('hamlet')
+        users = [
+            self.example_user('iago'),
+            self.example_user('cordelia'),
+        ]
+        all_users = users + [hamlet]
+        for user in all_users:
+            self.subscribe(user, stream)
+
+        do_soft_deactivate_users(users)
+        for user in users:
+            self.assertTrue(user.long_term_idle)
+
+        message_id = self.send_stream_message(hamlet.email, stream, 'Hello world!')
+        already_received = UserMessage.objects.filter(message_id=message_id).count()
+        do_catch_up_soft_deactivated_users(users)
+
+        for user in users:
+            user.refresh_from_db()
+            self.assertTrue(user.long_term_idle)
+            catch_up_received = UserMessage.objects.filter(message_id=message_id).count()
+            self.assertEqual(already_received + len(users), catch_up_received)
```

Comment:
We should validate the value of `last_active_message_id` here as well, since bugs in management of that seem like a pretty plausible way this algorithm could go wrong.  (and I think it's possible this PR has a bug where we're not doing that properly).

Reply:
Yep, that's a good check to have!